### PR TITLE
[DebugInfo] Backporting multiple upstream commits supporting fortran array debuggability

### PR DESCRIPTION
**Pull request by alokkrsharma**
This pull request is for back-porting already committed patches to upstream LLVM.

https://reviews.llvm.org/D79306 ([DebugInfo] llvm rejects DWARF operator DW_OP_push_object_address)
https://reviews.llvm.org/D79592 ([DebugInfo] support for DW_AT_data_location in llvm)
https://reviews.llvm.org/D80197 ([DebugInfo] Upgrade DISubrange to support Fortran dynamic arrays)
https://reviews.llvm.org/D83544 ([DebugInfo] Support for DW_AT_associated and DW_AT_allocated.)
https://reviews.llvm.org/D87500 ([DebugInfo] DISubrange support for fortran assumed size array)
https://reviews.llvm.org/D89141 [DebugInfo] Support for DWARF attribute DW_AT_rank
https://reviews.llvm.org/D89208 [DebugInfo] Support for DWARF operator DW_OP_over
https://reviews.llvm.org/D89218 [DebugInfo] Support for DW_TAG_generic_subrange

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
@alokkrsharma CI is enabled (make check-llvm, make check-clang) for this repository. Can you rebase and check if it passes?

**Comment by alokkrsharma** [comment]
> 
> 
> @alokkrsharma CI is enabled (make check-llvm, make check-clang) for this repository. Can you rebase and check if it passes?

Thanks @kiranchandramohan . It could be run. I can see one failure in LLVM Test (ubuntu-latest) with below failure message.

go build runtime/cgo: invalid flag in go:cgo_ldflag: -Wl,-allow-shlib-undefined

Primarily it does not look like due to my changes. Can I ignore this ?

**Comment by kiranchandramohan** [comment]
I believe the following three are in llvm-11, rest will come in llvm-12. Is that correct?
https://reviews.llvm.org/D79306 ([DebugInfo] llvm rejects DWARF operator DW_OP_push_object_address)
https://reviews.llvm.org/D79592 ([DebugInfo] support for DW_AT_data_location in llvm)
https://reviews.llvm.org/D80197 ([DebugInfo] Upgrade DISubrange to support Fortran dynamic arrays)

**Comment by alokkrsharma** [comment]
> 
> 
> I believe the following three are in llvm-11, rest will come in llvm-12. Is that correct?
> https://reviews.llvm.org/D79306 ([DebugInfo] llvm rejects DWARF operator DW_OP_push_object_address)
> https://reviews.llvm.org/D79592 ([DebugInfo] support for DW_AT_data_location in llvm)
> https://reviews.llvm.org/D80197 ([DebugInfo] Upgrade DISubrange to support Fortran dynamic arrays)

Yes, Correct.

**Comment by kiranchandramohan** [comment]
Which ones are the flang PRs that should go along with these? Would those be suitable candidates for consideration during the next two weeks? @alokkrsharma @shivaramaarao

**Comment by alokkrsharma** [comment]
> 
> 
> Which ones are the flang PRs that should go along with these? Would those be suitable candidates for consideration during the next two weeks? @alokkrsharma @shivaramaarao

The sequence should be 
1. Current one should go first
https://github.com/flang-compiler/classic-flang-llvm-project/pull/8
then port to LLVM 9.
2. https://github.com/flang-compiler/llvm/pull/86
Once LLVM side changes are in, we can move to Flang PRs
3. https://github.com/flang-compiler/flang/pull/901
4. https://github.com/flang-compiler/flang/pull/902
5. https://github.com/flang-compiler/flang/pull/913
6. https://github.com/flang-compiler/flang/pull/925
7. https://github.com/flang-compiler/flang/pull/926
8. https://github.com/flang-compiler/flang/pull/952

Please note that 3 to 8 each pull request is super set of previous one so after (2) we can directly jump to (8) which contains all commits from 3 to 8.

**Comment by bryanpkc** [comment]
Once these patches are merged into classic-flang-llvm-project/release_100, can we update ClassicFlang.cpp in release_100 to inform flang1/flang2 that the LLVM version has support for new (post-11.0) debuginfo?

**Review by gklimowicz — APPROVED** [review]
This passes our tests on OpenPOWER for LLVM 10.

**Comment by alokkrsharma** [comment]
> 
> 
> This passes our tests on OpenPOWER for LLVM 10.

Thanks for approving this. Can you please also have a look at https://github.com/flang-compiler/llvm/pull/86

**Comment by gklimowicz** [comment]
This passed our tests alone, but when I combine it with the changes in #952 (#901, #902, #913, #925, #926), I get failures in LLVM 10 builds. I'm going to backtrack and look at those individually with LLVM 10 and OpenPOWER.

For LLVM 9, I'm going to look at flang-compiler/llvm#86 with #952, and perhaps the individual PRs if that has failures. Stay tuned.

(I haven't tried testing against x86 or Arm.)

**Comment by bryanpkc** [comment]
@alokkrsharma Let me rephrase my question. Currently, the release_100 branch still pretends to be LLVM 9:

https://github.com/flang-compiler/classic-flang-llvm-project/blob/release_100/clang/lib/Driver/ToolChains/ClassicFlang.cpp#L958

This causes Flang to emit DebugInfo metadata as if the LLVM 9 is being used. After these patches are merged, the driver may need another command-line option to tell Flang that the LLVM being used has the latest DebugInfo patches applied. The same problem affects release_90 as well.

**Comment by kiranchandramohan** [comment]
Do you know where we set the arrays in fortran to column-major? Is it in Flang or LLVM?

**Comment by alokkrsharma** [comment]
> 
> 
> Do you know where we set the arrays in fortran to column-major? Is it in Flang or LLVM?

AFAIK, Flang and LLVM dont set it and debugger decides it based on DW_AT_language attribute (DW_ORD_col_major for fortran and row major for c/c++).
Reference: http://www.dwarfstd.org/doc/DWARF5.pdf (section 5.5)
If no ordering attribute is present, the default ordering for the source language
(which is indicated by the DW_AT_language attribute of the enclosing
compilation unit entry) is assumed.

**Comment by alokkrsharma** [comment]
> 
> 
> @alokkrsharma Let me rephrase my question. Currently, the release_100 branch still pretends to be LLVM 9:
> 
> https://github.com/flang-compiler/classic-flang-llvm-project/blob/release_100/clang/lib/Driver/ToolChains/ClassicFlang.cpp#L958
> 
> This causes Flang to emit DebugInfo metadata as if the LLVM 9 is being used. After these patches are merged, the driver may need another command-line option to tell Flang that the LLVM being used has the latest DebugInfo patches applied. The same problem affects release_90 as well.

Thanks, I was not aware of this. Though our planned order of merges should take care of this but it would be good to implement the option you are suggesting.
Our planned order of merges is
a) port to LLVM 10
b) port to LLVM 9
c) Flang changes assuming new debug info is available LLVM 9 onward

**Comment by alokkrsharma** [comment]
> 
> 
> This passed our tests alone, but when I combine it with the changes in #952 (#901, #902, #913, #925, #926), I get failures in LLVM 10 builds. I'm going to backtrack and look at those individually with LLVM 10 and OpenPOWER.
> 
> For LLVM 9, I'm going to look at [flang-compiler/llvm#86](https://github.com/flang-compiler/llvm/pull/86) with #952, and perhaps the individual PRs if that has failures. Stay tuned.
> 
> (I haven't tried testing against x86 or Arm.)

I have fixed the issue with Flang side of changes (build with FLANG_LLVM_EXTENSIONS), can you please try the build now?

**Comment by alokkrsharma** [comment]
Hi @kiranchandramohan and @bryanpkc,

Do you have any comment on this ?

**Review by bryanpkc — APPROVED** [review]
This built fine and passed tests for us.

**Comment by bryanpkc** [comment]
Will you submit another PR to remove the DIFortranSubrange code?

**Comment by alokkrsharma** [comment]
> 
> 
> Will you submit another PR to remove the DIFortranSubrange code?

Sure. I'll do that.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Comment by alokkrsharma** [comment]
> 
> 
> LGTM.

Thanks @kiranchandramohan . There was an unrelated test failure. Which is blocking the merge. Any suggestion ?

**Comment by kiranchandramohan** [comment]
@alokkrsharma I have merged this now. Will create a separate PR to disable the go bindings test as discussed in the previous call.

**Comment by alokkrsharma** [comment]
> 
> 
> Will you submit another PR to remove the DIFortranSubrange code?

I have raised PR https://github.com/flang-compiler/classic-flang-llvm-project/pull/10 for this. Should I raise one for release_90 as well ?
